### PR TITLE
support EC2 AWS credentials and IAM roles chain

### PR DESCRIPTION
**Pull request by fredericfran-gds**
In order to avoid passing explicitly the AWS credentials, the resource
has been modified to support retrieval of AWS credentials via EC2
metadata. In addition, IAM roles chain is now supported to provide the
flexibility required for GOV.UK scenario where the ec2 instances have
the credentials of the Concourse worker role which needs to be used to
assume the Concourse deployer role. The deployer role is then used to
assume a role in the GOV.UK production ECR to pull images.

To build and push image, e.g. procedure:

```sh
docker build -t registry-image-resource --target tests -f dockerfiles/alpine/Dockerfile .
docker tag registry-image-resource:latest fredericfran/registry-image:0.0.2
docker push fredericfran/registry-image:0.0.2
```

### CODE REVIEW

File: README.md
Comment (sengi) [inline README.md]: Missing a word?

```suggestion
  If set, these roles will be assumed in the specified order before authenticating to ECR.
```

File: types.go
Comment (sengi) [inline types.go]:
```suggestion
	var mySession *session.Session
```

File: types.go
Comment (sengi) [inline types.go]: Just iterate over a slice rather than branching inside the loop. Something like this:

```suggestion
		for roleIndex, roleArn := range (awsRoleArnsList[:len(awsRoleArnsList)-1]) {
			mySession = session.Must(session.NewSession(&aws.Config{
				Region:      aws.String(source.AwsRegion),
				Credentials: stscreds.NewCredentials(mySession, roleArn),
			}))
		}
		lastAwsRoleArn = awsRoleArnsList[len(awsRoleArnsList)-1]
		config = aws.Config{Credentials: stscreds.NewCredentials(mySession, lastAwsRoleArn)}
```

Thinking more broadly, this is probably a smell that there's a semantic distinction between the last ARN in the list and the others, which perhaps should be represented as two different config items.

File: README.md
Comment (fredericfran-gds) [reply README.md]: yes, I will fix.

File: types.go
Comment (fredericfran-gds) [reply types.go]: I will fix.

File: types.go
Comment (fredericfran-gds) [reply types.go]: Yeah, this is me wanting not to change the code too much, I think this can be fixed by creating new sessions in the for loop always and then use the session directly in `ecr.New` without additional config.

File: types.go
Comment (bilbof) [inline types.go:67]: We'll probably need to make this backwards compatible in order to upstream this. You could continue to accept `aws_role_arn` and transform it into `aws_role_arns`, and print out a warning. Your call if you want to do this, though it would be great if we didn't need to maintain this fork.

File: types.go
Comment (bilbof) [inline types.go:288]: Just to make this explicit and consistent with the rest of the file.

```suggestion
	if source.AwsEC2Credentials != "" {
```

File: commands/check.go
Comment (bilbof) [inline commands/check.go]: It looks like `source.AwsRegion` is required for both auth methods:

```suggestion
	    ((req.Source.AwsAccessKeyId != "" && req.Source.AwsSecretAccessKey != "") || (req.Source.AwsEC2Credentials != "")) && source.AwsRegion != "" {
```

**Review by bilbof — APPROVED** [review]
LGTM, I raised a couple of minor issues it'd be good to address.

Are there tests we can add for this change?

File: types.go
Comment (fredericfran-gds) [reply types.go:67]: yeah, I have been thinking about it. Where will we print the warning? Most resources do not print details except in debug mode?

File: types.go
Comment (fredericfran-gds) [reply types.go:288]: source.AwsEC2Credentials is a boolean, not string.

File: types.go
Comment (bilbof) [reply types.go:67]: Ideally we should handle this at the earliest point that `aws_role_arn` is provided, so perhaps in `types.go`.

File: commands/check.go
Comment (fredericfran-gds) [reply commands/check.go]: Good find, I will add that.

File: README.md
Comment (bilbof) [inline README.md]:
```suggestion
* `aws_role_arns`: *Optional. Default `""`.* A comma-delimited list of AWS IAM roles.
```

**Review by bilbof — APPROVED** [review]
LGTM, thanks for working on this!

It might be a bit surprising that `aws_role_arn` can now be multiple ARNs, and it would be great to have tests.

Perhaps you could open a PR to upstream these changes and see what the maintainers think?

**Pull request merged**
